Add tuning sliders

IvoFPV opened the pull request:
This is the second part of the #1584 which was separated. Now when that is merged we can start a discussion on this.
Adds tuning sliders that make tuning easier by providing a simple input in form of a slider which scales defaults up or down.

Filter sliders scales gyro and dterm filtering 0.5x-2x the default values. This has been found by @ctzsnooze to be a good maximum value without risking flyaways, but still requires a clean quad.

PID tuning sliders change P/D ratio, PD gain and Response. By using good helpicon messages I think we can really simplify the tuning process for beginners.

There is still work to do on the basic functionality of the sliders and looks, but the main part is discussing, testing and agreeing on safe values that won't cause issues. This may delay this for BF4.2 as we previously said to allow for testing.

UPDATE: We moved away from this being focused only on beginners and changed to having full range of motion on the sliders so everyone can use them. Values calculated by sliders are not restricted to keep them safe just like the normal manual inputs aren't. Filtering sliders are restricted to only 2x default values.

![image](https://user-images.githubusercontent.com/50072760/63766153-d8566380-c8ca-11e9-97a7-8b2cab7f3998.png)
![image](https://user-images.githubusercontent.com/50072760/63766170-e1dfcb80-c8ca-11e9-82a5-64f23a279229.png)

McGiverGim commented:
This will need a lot of testing to give them a good usability, but I think is a good idea and maybe the way to go. Good job.
It will be good if a user not expert can tune the quad only with the sliders. My first concern is that for achieve this target, we will need some other slider to differentiate between roll and pitch, maybe some multiplier/ratio for roll/pitch. I'm not an expert on tuning so maybe @ctzsnooze or @spatzengr can confirm if this is possible.

IvoFPV commented:
@McGiverGim There has been some improvement, could you take a look now? I would like your opinion on the styling and layout. Roll/pitch ratio sounds like a good idea, but I will have to discuss with others.

mikeller commented:
@McGiverGim: The agreement with @IvoFPV is that we'd add this _after_ the release to go with Betaflight 4.1, to give us and users enough time to test this with all sorts of different hardware configurations.

McGiverGim commented:
@mikeller I suppose you want to say "to go with Betaflight 4.2". And yes, that is what I understood in the other PR. This will go in the next release after Betaflight 4.1.

IvoFPV commented:
Actually since we moved away from this being a feature intended to be used by beginners only and gone to a different approach of having these available and useful for everyone we found that there isn't that much testing required in a sense of testing the slider calculated values on different builds. Just like the normal manual inputs aren't restricted for anyone, sliders aren't as well, providing a full range of adjustment. Going from super low values to maximum. 

Filtering sliders are restricted and we are trying to make sure their usefulness by testing the values on different builds, especially those like 6inch, micros.

I think we could put this together for 4.1 even. Because if sliders just offer an another way of moving the values manually, i.e. with no restrictions, we just have to test and confirm sliders functionality and consistency of the calculated values. And not really test every slider position. Also by keeping filtering sliders under a safe limit and having good helpicon messages, filter sliders are going to be the same.

Not trying to push something unsafe for 4.1, just want to get your thoughts on this approach.

bizmar commented:
To clarify a little, the goal of the current implementation is to be a convenient UI to change PID and filter values. The settings are are capped at reasonable values, but they well not be safe for all quads... They are there to work from micro to xclass. They have better built in help that BF has ever had, imho. And if you select extreme values there's a warning. It's disabled by default.

If not for 4.1 could we do a out of cycle update to the Configurator after a month or two? Maybe put it into RC and see how that goes before we make a final decision. Remove it if bad, keep it if good..

IvoFPV commented:
@mikeller I think we are close to finishing this. Maybe just a few message changes or if we find something else to change in the next few days. Other than that it is completely working and functional.
What are the chances of having this for 4.1? It is late, but still a full month of RC should be enough to resolve any issues that might arise or if some serious flaws are found it can easily be removed for release and reworked for 4.2. Maybe @ctzsnooze, @spatzengr can give their thoughts here on this and how it would improve the tuning process.

mikeller commented:
@IvoFPV: I thought we had an agreement that this would go into the release *after* the one to go with 4.1?

IvoFPV commented:
@mikeller We had, but that was because of the previous version of this was made to focus on beginners meaning that "safe values" would have to be thoroughly tested on many builds and it was best to leave it for 4.2 so there is enough time to do that. That was my first intention when I opened this PR. But things have changed since.

Now we moved to the different approach where actual testing of all possible combinations and values is not required. There are no "safe limits" on values like before, there are some which are there because of the scaling of maximum 2x we chose to use. So this just provides another way of tuning. Safety of the maximum PID values have never been tested before so the point is they don't have to be tested with sliders. Anyone can manually set unsafe values, for example 200 D term gains, why shouldn't they be able to do that with sliders.

Again, not trying to push this in any way, but saying this just to let you know that there is a chance it could go with the 4.1 release with this new approach and slider functionality tested during this month of RC phase.

ctzsnooze commented:
I apologise @mikeller for pushing for this in 4.1, but having played with it, I think it is a super valuable feature.  
For the average guy it will mean a great deal.  
I think it will save us a lot of support headaches that have arisen from people making stupid filter assignments.  It gives an easy way for tweakers to do so in a sensible manner, and see how much they are diverging from normal.  I think they will be far less likely to stuff up with sliders than ever before.  
And the filter sliders work very, very well with rpm filtering enabled, greatly simplifying how to adjust the filters upwards once the rpm filter cuts noise.
In that sense this is the perfect complement to the new ff and rpm features in 4.1.

ctzsnooze commented:
For instance, super simple to make a set of 6S race filters
![Screen Shot 2019-08-31 at 12 02 14](https://user-images.githubusercontent.com/11737748/64058061-d3562480-cbe8-11e9-869c-5039e4d2cbb3.jpg)

ctzsnooze commented:
And to cut my lowpass filter delay by 50% when using rpm filtering
![Screen Shot 2019-08-31 at 12 17 43](https://user-images.githubusercontent.com/11737748/64058098-63946980-cbe9-11e9-9673-252dea7b0ae6.jpg)

quadwizard commented:
Hey all,

This PR is such a great idea, the ability to very quickly visually see were you are with repsect to the defaults, quickly setup a new quad or default a single profile is so great.. 

This is really going to make it easier for "new users" and "keyboard warriors" to learn about tuning as its visual; the fact you can see the PIDs as they change and mannually tweek them if required; its going to save a lot of users a lot of dramas.. 

I really think this is: "one small step for Betaflight but one giant leap for its users"

It would be a welcome addition to 4.1 IMO as its going to fix a lot of problems that users cause themselves!

mikeller commented:
@IvoFPV, @ctzsnooze: I never doubted that the sliders would be super useful if they work for users in the way we want them to. My concerns are more with making sure that using the sliders will result in workable combinations of parameters for a reasonable percentage of the builds that users actually fly.
Before this we've never offered a solution that allowed users to tune parameters as a whole but only ever had them change them one by one, so we don't have much user feedback on how well these combined settings are working for the builds that users are actually using them for.
This is why I think it would be better if we gave them some more exposure with testers before we rolled them into a release.
Not sure yet if we are going to do a release candidate for configurator to make sure that the distribution of Unified Targets works, but if we do this will be an opportunity to give the sliders some exposure, and we can always roll them back if we get too much feedback from users with builds the sliders don't work for.

bizmar commented:
Could they be temporally merged now (if the contributing devs agree it's are ok to share), so we can ask for feedback on facebook when RC1 is released? We need "regular" folks. And a Jenkins link will go along way towords getting testers. I would rather now be distributing "one off" home-brew versions..

I only have two things that I would like addressing, the sliders are enabled by clicking on them whey they are disabled, in the PID tab that represents a very big "accidentally reset your pids" area. I would advise that enabling them be a button. The possibility of accidentally reseting pids when dragging the mouse with a trackpad, to change .. Antigravity.. for example, not noticing and saving, is too big in my opinion. 

And secondly what @quadwizard said in slack.. clarify filter strength naming, so it's more understandable which way you have to move the slider for more or less filtering.

Other bad choices will only become clear once it actually gets used. I tuned two quads this way now. They both needed very slight changes.. 

I showed it to my local group, most complaints were about splitting off rates, they all PID liked the sliders. And really really liked the filter slider..

IvoFPV commented:
@bizmar 
> I showed it to my local group, most complaints were about splitting off rates,
This was expected, layout change breaks the habit, but it is now much better in terms of having PID and Rateprofile settings in different tabs.
> And secondly what @quadwizard said in slack.. clarify filter strength naming, so it's more understandable which way you have to move the slider for more or less filtering.
We've updated them, labels should now make more sense.

@mikeller Yeah, I think RC for configurator would be a good idea in general. If it does not happen, maybe we could merge this and try to get feedback from guys running Jenkins builds and if we find out sliders have some flaws just pull them out for release, and rework them for later.
BTW, if we don't do RCs for configurator, users of BF 4.1 RC version are expected to run Jenkins builds of configurator anyways, right?

mikeller commented:
@bizmar:

> Could they be temporally merged now 

Can you explain what you mean by 'temporally merged'? 

> I showed it to my local group, most complaints were about splitting off rates, they all liked the PID sliders. And really really liked the filter slider..

How was the feedback after the users from your local group used it to tune their craft? For how many of them did the configuration they got from just using the sliders work out of the box?

@IvoFPV:

> Yeah, I think RC for configurator would be a good idea in general.

We don't normally do them as they are a lot of work, and the feedback we got from users early on after we switched to NW.js was that they hated to have to download and install new apps every couple of weeks.

> BTW, if we don't do RCs for configurator, users of BF 4.1 RC version are expected to run Jenkins builds of configurator anyways, right?

We normally try to get the new configurator out early-ish in the RC phase. This time I can't see this happening, as there are still a bunch of things that need to be done to make Unified Targets and VTX tables work the way they need to work - but that applies to a configurator RC as well, as this will be specifically to verify that the Unified Target distribution is holding up.

But yes, I guess we can always roll this back if we get too many reports from users that end up with unflyable configurations.

mikeller commented on src/js/fc.js:
```diff
@@ -538,6 +539,12 @@ var FC = {
         };
+
+        DEFAULT_PIDS = [
+            42, 60, 35, 20, 70,
```
Having defaults in here is somewhat problematic - these will have to reflect whatever is in the firmware, like for example these will have to be updated once betaflight/betaflight#8783 and betaflight/betaflight#8785 have been merged.

IvoFPV replied:
Yes, it will need to updated, but for now I guess this is the best option. We should have MSP support for default values for 4.2 possibly. That will also fix current need to change filter defaults which have the same issue.

mikeller replied:
Yes, automatically generated messages / client stubs for configuration over MSP would be a nice win, and a worthy target now that we are getting Unified Targets out of the way.

mikeller commented on src/js/tabs/pid_tuning.js:
```diff
@@ -408,20 +404,16 @@ TABS.pid_tuning.initialize = function (callback) {
-                    $('#pid_main .pid_titlebar2 th').attr('colspan', 6);
-                    $('.helpicon[i18n_title="pidTuningPidTuningTipFeedforward"]').hide();
-                    $('.helpicon[i18n_title="pidTuningPidTuningTipDMin"]').show();
+                    //$('#pid_main tr :nth-child(5)').show();
```
Can you please remove commented out code, or else add a comment stating why this was commented out?

IvoFPV replied:
It was to let current testers test out D_min OFF/ON, so the values are shown instead of hiding the whole D_min column. This will be uncommented for final version.

mikeller replied:
This seems to be getting closer to being mergeable, so these should be uncommented now I guess.

mikeller commented on src/js/tabs/pid_tuning.js:
```diff
@@ -1415,6 +1407,482 @@ TABS.pid_tuning.initialize = function (callback) {
+                }
+            }
+
+            function calculateNewPids() {
```
`pid_tuning.js` is a behemoth already. I think it would be good to split the actual slider based calculations into a separate file.
Thoughts?

IvoFPV replied:
Good point. Should I go ahead and do that?

mikeller replied:
Yes please.

IvoFPV replied:
Done.

IvoFPV commented:
With slider stuff now moved to a separate file, only thing left to do is to change default PIDs and filters after new settings get merged in the firmware.

IvoFPV commented:
@mikeller Yeah, I will uncomment those. I'm going to push some small fixes and update the defaults, so this will be ready tomorrow.

bizmar commented:
@mikeller 
> Can you explain what you mean by 'temporally merged'?
This was probably explained by the time I get to answering,  just in case.. Merge it now, so we can link to a Jenkins build for download (and some other source for Mac I guess). During RC we push a few tweaks based on feedback. Then toward the end of the RC period (if no "show-stopping" problems are found that would require us to pull it before) we evaluate the efficacy of the sliders based on feedback, and decide if we want them to stay or remove them for official release. Maybe if the problem is minor and well understood.. just needs some dev time. We could bring them back in a 0.0.x release. We really need it broader testing in any case.. 

> How was the feedback after the users from your local group used it to tune their craft? For how many of them did the configuration they got from just using the sliders work out of the box?
This was just a me and a few friends flying at a local field, I don't have a Win binary to share yet. So I just showed it to them on my laptop. If we want a small test group I can maybe post it first to our local FB group with about 200 members for the whole country (small country). Or someone else from a English speaking country (managing feedback will be a lot easier) can do that for their local group, so we "dip our tows in" before we expose it to the official FB group with 10.000 members. 😅

IvoFPV commented:
@mikeller Updated with new defaults. This also takes care of betaflight/betaflight#8784.
It can be merged when betaflight/betaflight#8783 gets merged.

McGiverGim commented:
My only problem with this approach is that this is an easier method to tune the quad, specially for newbies, but it is only available in the Configurator. The user will need to take a laptop with him.
With this implemention, client side, it can't be available at the OSD menu or Lua. But I don't know if it has enough interest at this moment to implement it at the FC side.

mikeller commented:
@IvoFPV: Seeing that, as @bizmar says, there seems to be the opinion that getting this into a release, and then out of it again, is not associated with significant time overhead, and that this does not seem to have been tested at all (except for users looking at a screen), I have doubts that we should move away from the original plan to only merge this after the release for Betaflight 4.1 has gone out. Maybe @bizmar can put together and release a version of the configurator that includes this in the meantime?

quadwizard commented:
@mikeller I have distributed a win32.exe to some associates here in AU and await their feedback from this weekend play; once this comes in I can outline that feedback...

IvoFPV commented:
@mikeller I don't think @bizmar expressed his view on this as he wanted to. It is not expected to be pulled out of the release, that could happen **only** if some serious issues are found or if users just don't like the idea during the RC phase. Both of those are very unlikely.
Saying it has not been tested at all is incorrect. More than 20 people that I know have actually tried them out themselves and all had positive feedback, this is not including many others that just have just seen how it works and liked the idea. We've had multiple logs testing out different filtering settings on more exotic quads and found values to work great. I don't know if we should let users wait another 6 months for this until 4.2 comes out. Having a separate build made by one of us individually with this in will probably not get recognized, therefore most regular users will not have a chance to try this out.
@McGiverGim This being only in the Configurator is a limitation, but not a really a big one as you might think. It should be available in the OSD, I agree, but that is for later versions after and if this gets accepted in the community.

bizmar commented:
@IvoFPV clarified the intent of my comment very well. 

My proposal to limit the scope of inical release was so we wouldn't get flooded with feedback. Not because I thought it was untested or unsafe. It is. And it has better and more understandable contextual documentation then the previous version.

mikeller commented:
@bizmar: 

> My proposal to limit the scope of inical release was so we wouldn't get flooded with feedback.

If you don't want to get flooded with feedback then the right thing to do is to wait until after the next release before this is merged. I do not see why you seem to think that this is super urgent?

I'd much rather skip the next release with this, and then plan an interstitial configurator release a couple of months down the track, once we have had feedback on it, and made the improvements that will come out of the feedback.

> Not because I thought it was untested or unsafe.

Well, it is not so much a matter of what you think, but from what you say you have no feedback from users who have actually tested it, and can confirm that it is safe, and just as importantly, actually works for them.

IvoFPV commented:
@mikeller 
>Well, it is not so much a matter of what you think, but from what you say you have no feedback from users who have actually tested it, and can confirm that it is safe, and just as importantly, actually works for them.

He is saying just for his flight group that they haven't tested this, and was only shown to them. @bizmar himself has tested this. Again, this has been tested by others like I said in my previous post.

I'm just pointing out that you think that this hasn't been tested at all which is not correct. I assume you got this impression from Bizmar's previous post and that has caused to change your mind to delay this. Guys like @ctzsnooze, @spatzengr stand behind this, tests have been done to make sure it is safe to use by actually testing it, not just assuming the values will work on others quads.
I just don't want to get this delayed because of the wrong impression that's all.

If you really think, even if the safety of this was guaranteed, that this should be delayed and there is a good reason for it, then I respect the decision.

ctzsnooze commented:
At present, testing Configurator from a PR is not at all easy to do.  Of the many pilots that I know well, only two can and do build Betaflight themselves, and none can build configurator.  

For user feedback, the proverbial *someone else* has to make binaries and vouch that they are safe and won't trash the tester's quads.  In practice there are only a handful of people in the whole world who would actually be in a position to test a change in Configurator code.  

Of those who have tested it to date, the perception is overwhelmingly positive.  Some aspects are not perfect, but the concept and execution is extremely good, and there will always be argument about user interface detail - no user interface will ever make everyone happy.  Overall my assessment is that this user interface is very approachable, clean and uncluttered, and is extremely usable both for beginners and experts.  

Frankly I think that something like this is long overdue.

There is a valid concern that a person may create flyaway settings by setting the sliders to stupid positions.  Well, of course, they can, and do, already, with numbers.  But now if they put all the sliders hard right into the red zone, and ignore the warnings that are very clear and obvious, they will know for sure that what they are doing is unwise.  And that is far better than the absence of any warning or any visual indication of potentially unwise values, which is the situation we are in without sliders.

Can the sliders be used to quickly create useful configurations that span the range of typical user cases?  In a word, yes.  I have been able to quickly generate setups that I previously knew worked well from my whoops to my 7" through 3" and 6S racers.  And it is really great to visually see how each 'looks'.  The value in having the same scale factor for each slider and the visual representation of the overall state of tune compared to defaults is very, very powerful.

Already we / I field many support problems from people who have randomly disabled filters or moved P and D in strange ways - with the 'one number at a time' method.  Typically they change numbers, don't know what they've done, don't know how to get everything back to 'normal', and have no visual reference as to how extreme their changes are.  I predict that having sliders will significantly reduce the amount of support required and meaningfully simplify the user experience, for tuning PIDs and in particular for modifying filters. 

The idea of simply scaling filters by a multiplier has been used by many people who've gone to rpm code and followed that exact method from the rpm wiki.  We know that this works extremely well.  At present, they have to read the wiki and copy and paste snippets into the CLI.  Now they only have to move a slider.  

Given the difficulties getting binaries, I think that all the testing that can reasonably be done, has been done.  

I appreciate that real concerns that exist when such a significant change is proposed, but there will never be an easy way or an ideal time to make a change like this.  Personally I think that the best possible time is to coincide with a release like 4.1.  Having seen how good this is I really hope I don't have to wait for 4.2.   I am confident that @IvoFPV will be responsive and will maintain the code going forward.  

So yes I am 100% in favour of this, and the sooner the better.

mikeller commented:
@ctzsnooze: Do you have any test results from users using this to tune different hardware configurations that you can supply to support your words?

> For user feedback, the proverbial someone else has to make binaries and vouch that they are safe and won't trash the tester's quads. In practice there are only a handful of people in the whole world who would actually be in a position to test a change in Configurator code.

Have you ever thought of the fact that for any release **someone has to make binaries**?
Yes, it is easy for you to say 'it is too hard for us to make binaries, we'll just have someone else do a release with this, and if it turns out to result in a lot of unflyable configurations, we'll be responsive and expect that person to do more releases until we happen to get it right'.
If you suspect that these changes lead to testers quads being thrashed, then by no means should this be included in an official release any time soon.

ctzsnooze commented:
Having set up a build environment myself, made binaries, tested then extensively myself and shared them with a number of people to test, I think its fair to say that it did occur to me that "someone has to make binaries".   And I do think this PR has had quite a bit of testing and fine tuning, enough to confirm that it's OK.

If you want to make an unflyable condition right now its fairly trivial with the current configurator.  Just a few clicks would do it.  Just click the sliders on the filters tab - turn all the lowpass filtering off.  That's incredibly easy to do, and enough to smoke the motors on most quads.  The filters could alternately be changed to a very, very wide range of values.  And while you're at it turn D up to 200.  Oh and click the Dynamic Notch feature off, that won't be difficult.  All these things can be done very easily.  It's true that there is a warning while the filters are being turned off - and that's good - but it's clear that the current code permits the creation of all kinds of un-flyable configurations with very little effort.  And it happens a great deal.

The slider concept constrains people to a narrower range of values than the manual method we currently have.  The maximum range of any slider only 0.5 to 2 times the defaults, whereas the old user interface encouraged people to just turn things off, or allowed them to set values within huge ranges.  

In addition, using sliders makes it very clear that the user is pushing the boundaries by providing warnings and clear visual indicators whenever values more than twice defaults would result.   With the current numeric entry fields there is no simple visual feedback as to the relative balance of P to D, or the overall strength of the PIDs or filters, etc.

My view is that the slider based method will reduce, not increase, the number of problems we currently get from users applying ridiculous settings.

IvoFPV commented:
>If you suspect that these changes lead to testers quads being thrashed, then by no means should this be included in an official release any time soon.

The sliders are meant to be an extension to manual tuning. By having sliders you are still manually tuning and must make sure your values are set correctly for your build. These changes could only ''thrash testers quad'' if the values are miscalculated in the sliders code which has been tested and doesn't happen.
Users still have the full range of tuning with these sliders like they do when inputing values one by one that means that yes, some positions of the sliders will cause flyaways and unexpected behaviour just like the same things happen on current master of configurator if you push things up manually.

>In addition, using sliders makes it very clear that the user is pushing the boundaries by providing warnings and clear visual indicators whenever values more than twice defaults would result.
My view is that the slider based method will reduce, not increase, the number of problems we currently get from users applying ridiculous settings.

This is a really good point, there is no warning if you push PID values up too much, we only get warnings if disabling certain filters. This changes with sliders. Pushing sliders above certain point displays a warning and will discourage users to push values that high therefore decreasing the amount of issues we get because of users applying crazy values themselves because they don't know what they are actually doing. Especially when there is a lack of tuning tips in helpicon messages. This has also been greatly improved with this PR.
New tips are really good in my opinion and provide just enough information that you could learn to tune just by reading them.

>and if it turns out to result in a lot of unflyable configurations

Only safety issue I see with these changes is that if the slider calculation algorithm were to fail and set wrong numbers in the PID value fields. This is the most important part and was thoroughly tested and works as expected like I said.
The values it generates are just scaled per defaults and by having the scaling go from 0.5x-2x there is enough range to tune any quad, going from micros to 7" and it was tested that just with sliders you could get a really good tune on these more exotic and of course normal 5" quads in a short amount of time, avoiding all possible drawbacks of setting individual values manually we discussed above.

So you are still manually tuning but have a really good visual representation and are still expected not to push things just like before. Unflyable configurations have been and will still be users fault while trying to avoid those with warnings and good tuning tips.

atomgomba commented:
I tested this and I got the "Sliders are disabled coz values are different than defaults". Then I clicked on "Reset all profile values" and the message did not disappear immediately. Though it has disappeared after I have moved the visually disabled looking sliders.

IvoFPV commented:
@atomgomba Which fimware version are you running?

atomgomba commented:
Hm, it seems it is looking for different defaults. How does it determine that values are the default or not? When I click on "Reset all profile values", Feed forward gets reset to 70, but if I click on the slider (don't move it away from Default range, just generate a click event), it sets 90 as default.

@IvoFPV: 
```
# version
# Betaflight / OMNIBUSF4SD (OBSD) 4.1.0 Aug 26 2019 / 20:58:05 (2b1d25639) MSP API: 1.42
```

IvoFPV commented:
@atomgomba Defaults have been changed in the recent build. That's why you have this issue. Update your firmware to the latest build and that should solve it.

McGiverGim commented:
To me is a little strange to have some grayed sliders that are enabled only touching them (and they reset the PID values).
My thoughts, you don't need to implement them, maybe they are a bad idea but if you think some can be useful go for it:
- If we don't have the default values, what about add a button to the warning message with "Reset PID values" to do it with the button and not touching the sliders? In this way a grayed slider is a grayed slider and the user can't touch the sliders by error.
- Some users maybe prefer to tune using the old PID table and not the sliders. They will show gray always and this is not clean. What about a check "Show tuning sliders" or maybe "Show EXPERIMENTAL tuning sliders" until we are sure it works as it must? In this way it will be an optional feature.
- The default values can change with 4.1 while in maintenance (and in future versions). This is one of the reasons why I don't like to have them in the Configurator, maybe at some moment we need to add a RESET MSP command to the firmware.
- As I stated before, it will be great to have MSP commands to change the values and not do it within the Configurator. In this way this great feature will be available at the OSD and Lua scripts, that is what most people use when tuning at the field and it will adapt to any future firmware version. 

Only some ideas...

McGiverGim commented on locales/en/messages.json:
```diff
@@ -3112,6 +3128,72 @@
+        "message": "Stick Response Gain:"
+    },
+    "pidTuningMasterSliderHelp": {
+        "message": "Generally larger quads need higher PID gains due to having a lower power to weight ratio.<br>Smaller quads (micros) generally need lower PID gains due to a higher power to weight ratio.<br>Higher values are for lower authority quads. Lower values are for higher authority quads."
```
Add a double `<br>` to be aligned with the others help messages. It will be great if you can add some `description` fields to the messages to help the translators.

IvoFPV replied:
I will add descriptions. What do you mean by double `<br>`?

McGiverGim replied:
The other similar help messages of this PR use `<br><br>` inside the text to separate paragraphs. This text has only one `<br>`.
![image](https://user-images.githubusercontent.com/2673520/64249339-b9cc1a00-cf13-11e9-8f0d-e069c1bbc202.png)
![image](https://user-images.githubusercontent.com/2673520/64249358-c6507280-cf13-11e9-948b-fd5252e2e5de.png)

IvoFPV replied:
Yeah, I will do that. 👍

McGiverGim commented on src/js/TuningSliders.js:
```diff
@@ -0,0 +1,313 @@
+        $('.slidersDisabled').hide();
+        this.initPidSlidersPosition();
+    }
+    if ((PIDs[1][0] > 110 || PIDs[1][1] > 120 || PIDs[1][2] > 70 || ADVANCED_TUNING.dMinPitch > 40) && !this.pidSlidersUnavailable) {
```
This values are better defined like constants or similar.

IvoFPV replied:
Yes, will change that.

IvoFPV commented:
@McGiverGim 
>If we don't have the default values, what about add a button to the warning message with "Reset PID values" to do it with the button and not touching the sliders? In this way a grayed slider is a grayed slider and the user can't touch the sliders by error.

This is a good idea and will definitely be implemented as some have noted before that you could accidentally reset your values just by touching any of the sliders and with the surface area of them so big it chances are not that slim.
The button to enable them after they have been disabled by manually tuning should fix that for sure.

>As I stated before, it will be great to have MSP commands to change the values and not do it within the Configurator. In this way this great feature will be available at the OSD and Lua scripts, that is what most people use when tuning at the field and it will adapt to any future firmware version.

MSP stuff will be addressed later in 4.2, of course if we choose to have this now for 4.1.

McGiverGim commented on src/js/TuningSliders.js:
```diff
@@ -0,0 +1,313 @@
+        $('.pid_filter input[name="gyroLowpass2Frequency"]').val() != Math.round(this.FILTER_DEFAULT.gyro_lowpass2_hz * this.gyroFilterSliderValue) ||
+        $('.pid_filter select[name="gyroLowpass2Type"]').val() != this.FILTER_DEFAULT.gyro_lowpass2_type) {
+
+        $('.tuningFilterSliders .sliderLabels tr:first-child').css('opacity', 0.3);
```
Maybe is better to add a class dinamically with the CSS content? Adding CSS content in the JS part makes difficult to make changes to the style later.

IvoFPV replied:
Probably this will be refactored with the new button to enable sliders, so there won't be any need for this.

McGiverGim commented:
Another suggestion:
- It is a good idea to let the users "share" the configuration. For example: "Move the master slider to the value X". 

Now it can't be done, it has no numeric values. Maybe you can add inside the circle a number, that goes from for example -X (at minimum) to +X (at maximum). In this way users can share his config: "I have a PID tuning of +1, +2.5 and -1" for example.

IvoFPV commented:
@McGiverGim 
![image](https://user-images.githubusercontent.com/50072760/64336770-1136bc80-cfde-11e9-96f2-efe38fafbd55.png)

McGiverGim commented:
Right! Sorry, I didn't remember that...

Another question, can the PD Balance or PD Gain have a name less technical? If this is intended for newbie users, maybe things like "Propwash handling" or "Overshoot control" will help (if we can find a suitable name, of course). At the tooltip we can explain what modifies for advanced users, of course.

IvoFPV commented:
@McGiverGim It is not strictly for beginner users anymore. Previous versions were. @ctzsnooze can give his thoughts on the naming again, but this is what we figured out to be the best at that time.

McGiverGim commented:
Is not for beginners only, but I suppose beginners will use it (and will need some easy names), and advanced users will do it too (for this kind of people the name is not important).

And what about the balance between roll/pitch? Will this sliders work for freestyle quads that usually need more gain in the pitch axis? What is the correct way to tune this kind of quads with the sliders?

mikeller commented:
@IvoFPV:

> maybe things like "Propwash handling" or "Overshoot control" will help

Fully agreed there. Even if it is not _only_ for beginners any more, the naming should aim to be such that the least experienced users using it can understand it - you can always add the more technical term for it in a tooltip.

IvoFPV commented:
In discussion with others we found out that roll to pitch ratio is not that important as a slider and that default ratio is good enough for most. Of course some fine tuning will have to be done manually. Sliders alone should be able to make any quad fly great, not best, but "just" great.

On the naming note, "Stick Response Gain" is already non PIDF value related name, so I guess we should make all of them like that as you say. Maybe something different than your recommendation but let's see what @ctzsnooze says.

ctzsnooze commented:
The only name I could think that we might easily change is:
`Stick Response Gain` -> `Stick Responsiveness`
The other two are fairly technical and not easily translated into a performance related name.  
For instance, if  P:D balance favours P,  we may slightly improve responsiveness, but get P oscillations.  If P:D balance favours D, it may reduce P oscillations and overshoot, but may increase noise and reduce responsiveness.  This is a complex set of possible outcomes that can't easily be put into one word.  I would prefer a name that would discourage a new user from playing with this slider.
For PD gain, usually in a quad with low authority, the master slider would be used to increase all the PIDs, but some people only want I and FF to go up, so they move the master up and reduce the PD gain.  This is a bit complicated again.  Higher PD gain will perhaps make the quad more 'locked in' and very low PD gain may make it loose, but also low PD gain may work better for 6S quads.  Again I can't see any good name for this, and would prefer the beginner to leave it alone.
To me P:D balance and PD gain are sort of expert/advanced sliders, not for beginners.
To be honest this is the names and order that I can best come up with.  
`Master multiplier`
`Stick responsiveness`
`PD gain (advanced)`
`P:D balance (advanced)`
I just can't figure any better ideas for names at the moment.

McGiverGim commented:
If we are using easy sliders in place that a complex PID table, I think we must give then names related to flight dynamics. 
Too much or less from any of them can produce problems, this is explained in the tooltip, but if you move some of them, is because you want to fix or change some of the flights characteristics.

P D balance can be authority for example. It can produce wobbles if too much authority but for this reason we have the red segment and a tooltip explain it.
P D ratio is clear that fixes overshoot and propwash. It can reduce responsiveness but this can be commented on the tooltip again.

If we can't find friendly descriptions of what can do any slider, then maybe we have chosen the wrong parameters to modify.

IvoFPV commented:
@McGiverGim The parameters they change are not wrong and if we can't explain a slider in one or two words that doesn't mean they shouldn't be there. As @ctzsnooze says, it is hard to call PD ratio overshoot control, because it can changes other things as well. Maybe we could use some more generic term that would take both factors into account, but I can't think of any. For now all of them are well explained in the tooltip so actual name is not that big of a problem in my opinion.

McGiverGim commented:
All the sliders change other things as well. The Stick Response Gain can produce overshoot too if too high (now less with FF2.0) but we are calling it Stick Response and not FF Gain.

I understand that you are trying to give an accurate term of what it does. I prefer not to be too accurate and be more user friendly. 

The vast majority of users use the defaults without touching the PID values. What I want is to give tools for this users to touch in a safe manner, and they must see something that flying they can understand.

For example, seeing the explanation at your tooltips:
- Master Gain: it changes the Authority of the quad.
- PD Ratio: to me the principal is to control Overshoot.
- PD Gain: to me the principal is reduce Propwash.
- Stick Responsiveness: to me is ok as is.

Maybe is a lot of simplification.

If we want to show the "problems" that can produce too low or high sliders, maybe we can add it to them:
![image](https://user-images.githubusercontent.com/2673520/64421899-fb3eff80-d0a2-11e9-976d-dde3d093ffa8.png)

Again, this is only my opinion of what the sliders must do. Take the PID tuning to the people that usually fly defaults, but let the advanced users understand what they do too with the more advanced explanation in the tooltip.

spatzengr commented:
How about this:

![image](https://user-images.githubusercontent.com/25570978/64423933-afc22c80-d075-11e9-920a-999c0933fcc4.png)

spatzengr commented:
Another thing I've thought about to give someone new some guidance on PID tuning (the arrow or arrows -- take one both or none).

Welcome thoughts by others on the idea and not sure if possible or super difficult in Configurator coding.  To me, this is the process as you wan to get your overall PID response sorted; then your P/D ratio; then your P&D Strenght; and then finally sort out your FF gains for stick tracking.  To the right increases PID push generally, but the downside is hot motors and overshooting if you go too far.

The only issue with the horizontal arrow is micros and high power to weight 5" quads may need to start below the Default middle position.

![image](https://user-images.githubusercontent.com/25570978/64428444-b8b8fb00-d081-11e9-9575-f60ab7a01d5a.png)

quadwizard commented:
@spatzengr I like the way you think... its all so logical :)

spatzengr commented:
@quadwizard , please convince my wife of that.  She would disagree.  ;-)  LOL.

mikeller commented:
Thanks for putting in the effort to get this out to users and collect feedback, @ctzsnooze, @IvoFPV. The feedback that I've seen seems to be overwhelmingly that this works for the builds that the users tested it with, so I think we can add this to 10.6 and be confident that most users will find it usable for their builds.

IvoFPV commented:
I will take a look later this day and see with others about naming changes.

IvoFPV commented:
Rebased and updated note spacing as in #1623. This should go with or after that.
We decided to keep the current names. Someone will always prefer to call it differently but with the explanation in the tooltip being really good it should be alright even for total beginners.
So this should now be ready.

mikeller commented:
@IvoFPV: Sorry, this needs a rebase now.

IvoFPV commented:
Rebased again. Should be all good now.

McGiverGim commented:
> We decided to keep the current names. 

Of course, I accept the final decision ;)

But I think we are losing an opportunity. There are people that has afraid to touch the PID tab because they don't understand nothing about it. The don't touch any of them. If you know some people like this, there are a ton, ask them if they will touch a parameter with a name like "P/D Ratio" or "P/D Gain". They will say no, for them they are parameters or the PID that they don't touch. Ask him if he will touch if this parameter is called "Propwash Control", I'm sure at least he will read the tooltip.

For the rest, good work with this. At least for a first version it will be enough.

IvoFPV commented:
I understand but it is hard to do that without misrepresenting the slider as we discussed. Sliders should make people less afraid to make change even though it may not be the most beginner friendly slider name. I'm sure for guys who "teach" how to tune it will be easy to explain and will get those guys who were afraid to start tuning.
I really hope this gains traction and for 4.2 we do many things with this concept 👍

quadwizard commented:
I for one would love to see sliders for rates....

McGiverGim commented:
@IvoFPV one question, I know this has been merged...

Will the slider for filters work with the new RPM Filter? The theory is that with it, we must "reduce" the dyn filter, not only move the frequencies of the LPF filters.

spatzengr commented:
It is not a "must", but if you are adding a filter, you should reduce something to compensate *IF* all is looking good before the add.  Either higher slider values on lowpasses, or the RPM pastes narrowes the DN to Q=250 and Width = 0 (single notch) to compensate for the latency add of the RPM notches.

ctzsnooze commented:
We have advice in the tuning guide about changing the dynamic notch.
My personal view is that when rpm is enabled the dynamic notch should automatically change to the rpm type defaults.  That requires some 'auto' mode, however, since some people may want to further change the dynamic notch settings in rpm mode, so we need auto/manual mode and all the manual settings in the CLI.  This may come in future but is not now.  So maybe we can have information in the tooltip advising on changing the width and Q of the dynamic notch.
As for what about lowpass filtering, it's best to just enable rpm, fly and check its good, then move the sliders for filters up in a few steps, aiming to get to 2x unless motors get hot.

McGiverGim commented:
I thought the slider will modify the dynamic notch settings, not only the lpf.

IvoFPV commented:
@McGiverGim Since MEDIUM for dynamic notch range is the new default there is no need to adjust it with sliders. That would be an issue if it was still on AUTO. Or maybe you are thinking of changing notch Q with the slider?

McGiverGim commented:
True, if our recommendation is reduce to only one notch, maybe we must get to this when the slide is moved.

ghostface commented:
I'd suggest that the gyro filter slider indicates somehow that it is not working profile-dependant.
I first thought it was a bug until I saw uav techs video where he explained why this is.

hydra commented:
FYI: Tuning sliders UI always shows the 'Enable Sliders' button when `USE_DYN_LPF` is not enabled in the firmware, it's because `MSP_FILTER_CONFIG` outputs `0` for some of the settings and the UI code in `updateFilterSlidersDisplay` checks the values.  However, even if you enable the sliders, then save the config, then go back to the sliders view or refresh it the 'Enable Sliders' button will re-appear.
